controllers/machinedeployment_sync.go

```diff
@@ -406,55 +402,42 @@ func calculateStatus(allMSs []*clusterv1.MachineSet, newMS *clusterv1.MachineSet
 
 func (r *MachineDeploymentReconciler) scaleMachineSet(ctx context.Context, ms *clusterv1.MachineSet, newScale int32, deployment *clusterv1.MachineDeployment) error {
 	if ms.Spec.Replicas == nil {
-		return errors.Errorf("spec replicas for machine set %v is nil, this is unexpected", ms.Name)
+		return errors.Errorf("spec.replicas for MachineSet %v is nil, this is unexpected", client.ObjectKeyFromObject(ms))
 	}
 
-	// No need to scale
-	if *(ms.Spec.Replicas) == newScale {
-		return nil
-	}
-
-	var scalingOperation string
-	if *(ms.Spec.Replicas) < newScale {
-		scalingOperation = "up"
-	} else {
-		scalingOperation = "down"
-	}
-
-	return r.scaleMachineSetOperation(ctx, ms, newScale, deployment, scalingOperation)
-}
-
-func (r *MachineDeploymentReconciler) scaleMachineSetOperation(ctx context.Context, ms *clusterv1.MachineSet, newScale int32, deployment *clusterv1.MachineDeployment, scaleOperation string) error {
-	if ms.Spec.Replicas == nil {
-		return errors.Errorf("spec replicas for machine set %v is nil, this is unexpected", ms.Name)
-	}
-
-	sizeNeedsUpdate := *(ms.Spec.Replicas) != newScale
-
 	annotationsNeedUpdate := mdutil.ReplicasAnnotationsNeedUpdate(
 		ms,
 		*(deployment.Spec.Replicas),
 		*(deployment.Spec.Replicas)+mdutil.MaxSurge(*deployment),
 	)
 
-	if sizeNeedsUpdate || annotationsNeedUpdate {
-		patchHelper, err := patch.NewHelper(ms, r.Client)
-		if err != nil {
-			return err
-		}
+	// No need to scale nor setting annotations, return.
+	if *(ms.Spec.Replicas) == newScale && !annotationsNeedUpdate {
+		return nil
+	}
+
+	// If we're here, a scaling operation is required.
+	patchHelper, err := patch.NewHelper(ms, r.Client)
+	if err != nil {
+		return err
+	}
 
-		*(ms.Spec.Replicas) = newScale
-		mdutil.SetReplicasAnnotations(ms, *(deployment.Spec.Replicas), *(deployment.Spec.Replicas)+mdutil.MaxSurge(*deployment))
+	// Save original replicas to log in event.
+	originalReplicas := *(ms.Spec.Replicas)
```

Comment:
let's try to avoid pointer dereferencing without nil checks as much as possible as it could lead to a panic (I know there were a lot before this change, just a suggestion for further improvement)

Reply:
makes total sense, this one is checked here https://github.com/kubernetes-sigs/cluster-api/pull/4506/files#diff-c6d8dd16b2cc683f753e58776ca23da70729fe8bd2834ce7b2626f570c189f4cR404

Reply:
perfect, I missed that one - thanks

Reply:
I think there are a few others that need a check, like deployment.Spec.Replicas

Reply:
thanks! fixed and covered by a testCase.